---
title: "Using gatom package"
author: "Anstasiia Gainullina, Mariia Emelianova, Alexey Sergushichev"
date: "September 2023"
output:
  BiocStyle::html_document:
    toc: true
    toc_float: false
    self_contained: false
vignette: >
  %\VignetteIndexEntry{Using gatom package}
  %\VignetteEngine{knitr::rmarkdown}
  \VignetteEncoding{UTF-8}"
---

This tutorial describes an R-package for finding active metabolic modules in 
atom transition network.

# Installation

* You can install **gatom** via `BiocManager`:
```{R eval = FALSE}
# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# 
# BiocManager::install("gatom")
```


# Example workfow

In this example we will find an active metabolic module based on macrophage 
activation gene expression data.

```{r message=FALSE}
library(gatom)
library(data.table)
library(igraph)
library(mwcsr)
```

First let's load data with atom mappings (`network` object), enzyme annotations 
for mouse (`org.Mm.eg.gatom`) and metabolite annotations (`met.kegg.db.rda`):

```{r}
data("networkEx")
data("org.Mm.eg.gatom.annoEx")
data("met.kegg.dbEx")
```

Loading input data:

```{r message=FALSE}
data("met.de.rawEx")
data("gene.de.rawEx")
```

Getting metabolic graph (depending on `topology` parameter, the graph vertices 
can correspond to `atoms` or `metabolites`):

```{r}
g <- makeMetabolicGraph(network=networkEx, 
                        topology="atoms",
                        org.gatom.anno=org.Mm.eg.gatom.annoEx, 
                        gene.de=gene.de.rawEx,
                        met.db=met.kegg.dbEx, 
                        met.de=met.de.rawEx)
print(g)
```

Scoring graph, obtaining an instance of SGMWCS (Signal Generalized Maximum 
Weight Subgraph) problem instance:

```{r message=FALSE, warning=FALSE}
gs <- scoreGraph(g, k.gene=25, k.met=25)
```

Initialize an SMGWCS solver (a heuristic relax-and-cut solver `rnc_solver` 
is used for simplicity, check out `mwcsr` package documentation for more options, 
and see [Using exact solver](#exact-solver) section for recommended way):

```{r}
solver <- rnc_solver()
```

Finding a module:

```{r message=FALSE, warning=FALSE}
res <- solve_mwcsp(solver, gs)
m <- res$graph
```
    
```{r}
print(m)
head(E(m)$label)
head(V(m)$label)
```

The module can be plotted in R Viewer with `createShinyCyJSWidget()`:

```{r}
set.seed(42)
createShinyCyJSWidget(m)
```

Next, reactions without highly changing genes but with high average 
expression can be added.

```{r message=FALSE, warning=FALSE}
m.ext <- addHighlyExpressedEdges(m, gs)
createShinyCyJSWidget(m.ext)
```

Sometimes, as in example above, the same metabolite can appear multiple times 
in the module via different atoms. In such cases it is useful
to either connect atoms belonging to the same metabolite with edges 
or to collapse them into one vertex.

To connect atoms use `connectAtomsInsideMetabolite` function:

```{r message=FALSE, warning=FALSE}
m1 <- connectAtomsInsideMetabolite(m.ext)
createShinyCyJSWidget(m1)
```

And to collapse them into one vertex use `collapseAtomsIntoMetabolites`:

```{r message=FALSE, warning=FALSE}
m2 <- collapseAtomsIntoMetabolites(m.ext)
createShinyCyJSWidget(m2)
```


# Saving modules

We can save the module to graphml format with `write_graph()` function from `igraph`:

```{r}
write_graph(m, file="M0.vs.M1.graphml", format="graphml")
```

Or it can be saved to an interactive html widget:

```{r message=FALSE}
saveModuleToHtml(module = m, file = "M0.vs.M1.html", name="M0.vs.M1")
```

```{r message=FALSE, warning=FALSE}
htmltools::includeHTML("M0.vs.M1.html")
```


Or we can save the module to dot format:

```{r}
saveModuleToDot(m, file="M0.vs.M1.dot", name="M0.vs.M1")
```

Such dot file can be further used to generate svg file using `neato` tool 
from graphviz suite:

```{r eval=FALSE}
system("neato -Tsvg M0.vs.M1.dot > M0.vs.M1.svg", ignore.stderr=TRUE)
```


Alternatively, the module can be saved to pdf format with a nice layout. 
You may vary the meaning of repel force and the number of iterations of 
repel algorithm for label layout. 
Note, that the larger your graph is the softer force you should use. 
You may also set different seed for different variants of edge layout.

```{r results="hide", message=FALSE, warning=FALSE}
set.seed(42)
saveModuleToPdf(m, file="M0.vs.M1.pdf", name="M0.vs.M1", 
                n_iter=100, force=1e-5)
```


# Example on full data and full network

Full dataset for the example can be downloaded here:

```{r message=FALSE}
library(R.utils)
library(data.table)

met.de.raw <- fread("http://artyomovlab.wustl.edu/publications/supp_materials/GAM/Ctrl.vs.MandLPSandIFNg.met.de.tsv.gz")
gene.de.raw <- fread("http://artyomovlab.wustl.edu/publications/supp_materials/GAM/Ctrl.vs.MandLPSandIFNg.gene.de.tsv.gz")
```

Full pre-generated combined network (ref. [Networks](#networks) for details) 
and enzyme annotation can be downloaded from  [http://artyomovlab.wustl.edu/publications/supp_materials/GATOM/](http://artyomovlab.wustl.edu/publications/supp_materials/GATOM/):

```{r}
network.combined <- readRDS(url("http://artyomovlab.wustl.edu/publications/supp_materials/GATOM/network.combined.rds"))
met.combined.db <- readRDS(url("http://artyomovlab.wustl.edu/publications/supp_materials/GATOM/met.combined.db.rds"))

org.Mm.eg.gatom.anno <- readRDS(url("http://artyomovlab.wustl.edu/publications/supp_materials/GATOM/org.Mm.eg.gatom.anno.rds"))
```

For better work of the combined network we highly recommend using additional 
supplementary gene files (ref. [Supplementary Genes](#suppl-genes))

```{r}
gene2reaction.extra <- fread("http://artyomovlab.wustl.edu/publications/supp_materials/GATOM/gene2reaction.combined.mmu.eg.tsv", colClasses="character")
```

Running `gatom` on combined network and full dataset:

```{r}
cg <- makeMetabolicGraph(network=network.combined,
                         topology="atoms",
                         org.gatom.anno=org.Mm.eg.gatom.anno,
                         gene.de=gene.de.raw,
                         met.db=met.combined.db,
                         met.de=met.de.raw,
                         gene2reaction.extra=gene2reaction.extra)

cgs <- scoreGraph(cg, k.gene = 50, k.met = 50)

solver <- rnc_solver()
set.seed(42)
sol <- solve_mwcsp(solver, cgs)
cm <- sol$graph
cm
```

```{r message=FALSE, warning=FALSE}
createShinyCyJSWidget(cm)
```

# Networks {#networks}

We provide four types of networks that can be used for analysis:

1. KEGG network
2. Rhea network
3. Combined network
4. Rhea lipid subnetwork 

## KEGG

KEGG network consists of `network.kegg.rds` & `met.kegg.db.rds` files and is 
based on [KEGG database](https://www.genome.jp/kegg/kegg1.html). 

Both metabolites and reactions have KEGG IDs.

```{r}
network <- readRDS(url("http://artyomovlab.wustl.edu/publications/supp_materials/GATOM/network.kegg.rds"))
met.db <- readRDS(url("http://artyomovlab.wustl.edu/publications/supp_materials/GATOM/met.kegg.db.rds"))
```

Running `gatom` on this dataset:

```{r}
kg <- makeMetabolicGraph(network=network, 
                        topology="atoms",
                        org.gatom.anno=org.Mm.eg.gatom.anno, 
                        gene.de=gene.de.raw,
                        met.db=met.db, 
                        met.de=met.de.raw)

kgs <- scoreGraph(kg, k.gene = 50, k.met = 50)

solver <- rnc_solver()
set.seed(42)
sol <- solve_mwcsp(solver, kgs) 
km <- sol$graph
km
```

```{r message=FALSE, warning=FALSE}
createShinyCyJSWidget(km)
```

## Rhea

Rhea network consists of `network.rhea.rds` & `met.rhea.db.rds` files and 
is based on [Rhea database](https://www.rhea-db.org/). 

Reactions in Rhea have their own IDs, but unlike KEGG, metabolite IDs come 
from a separate database -- [ChEBI database](https://www.ebi.ac.uk/chebi/)

To use Rhea network download the following files:

```{r}
network.rhea <- readRDS(url("http://artyomovlab.wustl.edu/publications/supp_materials/GATOM/network.rhea.rds"))
met.rhea.db <- readRDS(url("http://artyomovlab.wustl.edu/publications/supp_materials/GATOM/met.rhea.db.rds"))
```

For proper work of the Rhea network we also need a corresponding 
supplementary gene file (ref. [Supplementary Genes](#suppl-genes))

```{r}
gene2reaction.extra <- fread("http://artyomovlab.wustl.edu/publications/supp_materials/GATOM/gene2reaction.rhea.mmu.eg.tsv", colClasses="character")
```

And run `gatom` on Rhea network:

```{r}
rg <- makeMetabolicGraph(network=network.rhea,
                         topology="atoms",
                         org.gatom.anno=org.Mm.eg.gatom.anno,
                         gene.de=gene.de.raw,
                         met.db=met.rhea.db,
                         met.de=met.de.raw,
                         gene2reaction.extra=gene2reaction.extra)

rgs <- scoreGraph(rg, k.gene = 50, k.met = 50)

solver <- rnc_solver()
set.seed(42)
sol <- solve_mwcsp(solver, rgs)
rm <- sol$graph
rm
```

```{r message=FALSE, warning=FALSE}
createShinyCyJSWidget(rm)
```

## Combined network

Combined network comprises not only KEGG and Rhea reactions, but also transport 
reactions from [BIGG database](http://bigg.ucsd.edu/).

This means that reactions in such network have either KEGG or Rhea or BIGG IDs, 
and metabolite IDs are KEGGs and ChEBIs.

## Rhea lipid subnetwork

Rhea lipid subnetwork is subset of Rhea reactions that involve lipids, 
and it consists of `network.rhea.lipids.rds` & `met.rhea.lipids.db.rds` files. 

```{r}
network.lipids <- readRDS(url("http://artyomovlab.wustl.edu/publications/supp_materials/GATOM/network.rhea.lipids.rds"))
met.lipids.db <- readRDS(url("http://artyomovlab.wustl.edu/publications/supp_materials/GATOM/met.lipids.db.rds"))
```

For proper work of the lipid network we will also need a corresponding 
supplementary gene file (ref. [Supplementary Genes](#suppl-genes))

```{r}
gene2reaction.extra <- fread("http://artyomovlab.wustl.edu/publications/supp_materials/GATOM/gene2reaction.rhea.mmu.eg.tsv", colClasses="character")
```

To test lipid network we will use a different dataset:

```{r}
met.de.lipids <- fread("https://artyomovlab.wustl.edu/publications/supp_materials/GATOM/Ctrl.vs.HighFat.lipid.de.csv")
```

For lipid network we recommend setting topology to `metabolites` 
(ref. [Using metabolite-level network](#met-topology)):

```{r}
lg <- makeMetabolicGraph(network=network.lipids,
                         topology="metabolites",
                         org.gatom.anno=org.Mm.eg.gatom.anno,
                         gene.de=NULL,
                         met.db=met.lipids.db,
                         met.de=met.de.lipids,
                         gene2reaction.extra=gene2reaction.extra)

lgs <- scoreGraph(lg, k.gene = NULL, k.met = 50)

solver <- rnc_solver()
set.seed(42)
sol <- solve_mwcsp(solver, lgs)
lm <- sol$graph
lm
```

```{r message=FALSE, warning=FALSE}
createShinyCyJSWidget(lm)
```

If IDs for DE table for metabolites are of type `Species` we can process 
metabolite labels into more readable ones:

```{r message=FALSE, warning=FALSE}
lm1 <- abbreviateLabels(lm, orig.names = TRUE, abbrev.names = TRUE)

createShinyCyJSWidget(lm1)
```


# Misc

## Supplementary gene files {#suppl-genes}

For combined, Rhea and lipid networks we provide supplementary files with genes 
that either come from proteome or are not linked to a specific enzyme. 
These files are organism-specific and are also available at  [http://artyomovlab.wustl.edu/publications/supp_materials/GATOM/](http://artyomovlab.wustl.edu/publications/supp_materials/GATOM/).

```{r}
network.combined <- readRDS(url("http://artyomovlab.wustl.edu/publications/supp_materials/GATOM/network.combined.rds"))
met.combined.db <- readRDS(url("http://artyomovlab.wustl.edu/publications/supp_materials/GATOM/met.combined.db.rds"))
gene2reaction.extra <- fread("http://artyomovlab.wustl.edu/publications/supp_materials/GATOM/gene2reaction.combined.mmu.eg.tsv", colClasses="character")

gg <- makeMetabolicGraph(network=network.combined, 
                         topology="atoms",
                         org.gatom.anno=org.Mm.eg.gatom.anno, 
                         gene.de=gene.de.raw,
                         met.db=met.combined.db, 
                         met.de=met.de.raw, 
                         gene2reaction.extra=gene2reaction.extra)
gg
```

## Non-enzymatic reactions

Optionally, we can also preserve non-enzymatic reactions that are found in the network. 
This can be done by setting `keepReactionsWithoutEnzymes` to `TRUE`:

```{r}
ge <- makeMetabolicGraph(network=network.combined, 
                         topology="atoms",
                         org.gatom.anno=org.Mm.eg.gatom.anno, 
                         gene.de=gene.de.raw,
                         met.db=met.combined.db, 
                         met.de=met.de.raw, 
                         gene2reaction.extra=gene2reaction.extra, 
                         keepReactionsWithoutEnzymes=TRUE)
ge
```

## Using exact solver {#exact-solver}

For proper analysis quality we recommend to use exact SGMWCS solver `virgo_solver()`, 
which requires Java (version >= 11) and CPLEX (version >= 12.7) to be installed. 
If the requirements are met you can then find a module as following:

```{r eval=FALSE}
vsolver <- virgo_solver(cplex_dir=Sys.getenv("CPLEX_HOME"), 
                        threads=4, penalty=0.001, log=1)
sol <- solve_mwcsp(vsolver, gs) 
m <- sol$graph
```

Edge penalty option there is used to remove excessive redundancy in genes.

## Running with no metabolite data

If there is no metabolite data in your experiment assign `met.de` and `k.met` to `NULL`:

```{r}
g <- makeMetabolicGraph(network=networkEx, 
                        topology="atoms",
                        org.gatom.anno=org.Mm.eg.gatom.annoEx, 
                        gene.de=gene.de.rawEx,
                        met.db=met.kegg.dbEx, 
                        met.de=NULL)
gs <- scoreGraph(g, k.gene = 50, k.met = NULL)
```

## Running with no gene data

If there is no gene data in your experiment assign `gene.de` and `k.gene` to `NULL`:

```{r}
g <- makeMetabolicGraph(network=networkEx, 
                        topology="atoms",
                        org.gatom.anno=org.Mm.eg.gatom.annoEx, 
                        gene.de=NULL,
                        met.db=met.kegg.dbEx, 
                        met.de=met.de.rawEx)
gs <- scoreGraph(g, k.gene = NULL, k.met = 50)
```


## Using metabolite-level network {#met-topology}

Sometimes it could make sense to work with metabolite-metabolite topology 
of the network, not atom-atom one. 
Such network is less structured, but contains more genes.

```{r}
gm <- makeMetabolicGraph(network=network, 
                        topology="metabolite",
                        org.gatom.anno=org.Mm.eg.gatom.anno, 
                        gene.de=gene.de.raw,
                        met.db=met.db, 
                        met.de=met.de.raw)

gms <- scoreGraph(gm, k.gene = 50, k.met = 50)

solver <- rnc_solver()
set.seed(42)
sol <- solve_mwcsp(solver, gms) 
mm <- sol$graph
mm
```

## Pathway annotation

To get functional annotation of obtained modules by KEGG and Reactome
metabolic pathways, we can use hypergeometric test with `fora()` function from `fgsea` package.

```{r}
foraRes <- fgsea::fora(pathways=org.Mm.eg.gatom.anno$pathways,
                       genes=E(km)$gene,
                       universe=unique(E(kg)$gene),
                       minSize=5)
foraRes[padj < 0.05]
```

Optionally, redundancy in pathways can be decreased with `collapsePathwaysORA()` function:

```{r}
mainPathways <- fgsea::collapsePathwaysORA(
  foraRes[padj < 0.05],
  pathways=org.Mm.eg.gatom.anno$pathways,
  genes=E(km)$gene,
  universe=unique(E(kg)$gene))
foraRes[pathway %in% mainPathways$mainPathways]
```


```{r}
sessionInfo()
```
