---
title: "Using gatom package"
author: "Anstasiia Gainullina, Alexey Sergushichev"
output:
  BiocStyle::html_document:
    toc_float: yes
vignette: >
  %\VignetteIndexEntry{Using gatom package}
  %\VignetteEngine{knitr::rmarkdown}
  \VignetteEncoding{UTF-8}"
---

This tutorial describes an R-package for finding active metabolic modules in atom transition network.

# Installation
 
 * **gatom** package depends on **mwcsr** for working with appearing maximum weight connected subgraph problem instances, which can be installed with `devtools`:
```{R eval = FALSE}
    #install.packages("devtools")
    devtools::install_github("ctlab/mwcsr")
```

* After that you can install **gatom** package via `devtools` as well:
```{R eval = FALSE}
    #install.packages("devtools")
    devtools::install_github("ctlab/gatom")
```


# Example workfow

In this example we will find an active metabolic module based on macrophage activation 
gene expression data.

```{r message=FALSE}
library(gatom)
library(data.table)
library(igraph)
library(mwcsr)
```

First let's load data with atom mappings (`network` object),
enzyme annotations for mouse (`org.Mm.eg.gatom`)
and metabolite annotations (`met.kegg.db.rda`):

```{r}
data("networkEx")
data("org.Mm.eg.gatom.annoEx")
data("met.kegg.dbEx")
```

Loading input data:

```{r message=FALSE}
data("met.de.rawEx")
data("gene.de.rawEx")
```

Getting metabolic graph (depending on `topology` parameter, the graph vertices can correspond to `atoms` or `metabolites`):

```{r}
g <- makeMetabolicGraph(network=networkEx, 
                        topology="atoms",
                        org.gatom.anno=org.Mm.eg.gatom.annoEx, 
                        gene.de=gene.de.rawEx,
                        met.db=met.kegg.dbEx, 
                        met.de=met.de.rawEx)
print(g)
```

Scoring graph, obtaining an instance of SGMWCS (Signal Generalized Maximum Weight Subgraph)
problem instance:

```{r message=FALSE, warning=FALSE}
gs <- scoreGraph(g, k.gene=25, k.met=25)
```

Initialize an SMGWCS solver (a heuristic relax-and-cut solver `rnc_solver` is used for simplicity, check out `mwcsr` package documentation for more options, and see "Using exact solver" section for recommended way):

```{r}
solver <- rnc_solver()
```

Finding a module:

```{r message=FALSE, warning=FALSE}
res <- solve_mwcsp(solver, gs)
m <- res$graph
```
    
```{r}
print(m)
head(E(m)$label)
head(V(m)$label)
```

We can save the module to dot format to generate pdf or svg file using `neato` tool from graphviz suite.

```{r}
saveModuleToDot(m, file="M0.vs.M1.dot", name="M0.vs.M1")
system("neato -Tsvg M0.vs.M1.dot > M0.vs.M1.svg", ignore.stderr=T)
```

```{r echo=FALSE}
knitr::include_graphics("M0.vs.M1.svg")
```

The module can also be saved to xgmml format and later opened in Cytoscape.

```{r}
saveModuleToXgmml(m, file="M0.vs.M1.xgmml", name="M0.vs.M1")
```

The module can also be saved to pdf format with a nice layout.
You may vary the meaning of repel force and the number of iterations of repel algorithm for label layout. Note, that the larger your graph is the softer force you should use. 
You may also change seed for different variants of edge layout.

```{r results="hide", message=FALSE, warning=FALSE}
saveModuleToPdf(m, file="M0.vs.M1.pdf", name="M0.vs.M1", n_iter=100, force=1e-5, seed=10)
```

```{r echo=FALSE, eval=FALSE, message=FALSE, warning=FALSE}
knitr::include_graphics("M0.vs.M1.pdf")
```

<img src="https://rawgit.com/ctlab/gatom/master/inst/M0.vs.M1.png"/>

Next, reactions without highly changing genes but with high average 
expression can be added.

```{r}
m.ext <- addHighlyExpressedEdges(m, gs)
saveModuleToDot(m.ext, file="M0.vs.M1.ext.dot", name="M0.vs.M1")
system("neato -Tsvg M0.vs.M1.ext.dot > M0.vs.M1.ext.svg", ignore.stderr = T)
```

```{r echo=FALSE}
knitr::include_graphics("M0.vs.M1.ext.svg")
```

Sometimes, as in example above, the same metabolite can appear multiple 
times in the module via different atoms. In such cases it is useful
to either connect atoms belonging to the same metabolite with edges
with `connectAtomsInsideMetabolite` function.

```{r}
m1 <- connectAtomsInsideMetabolite(m.ext)
saveModuleToDot(m1, file="M0.vs.M1.v1.dot", name="M0.vs.M1")
system("neato -Tsvg M0.vs.M1.v1.dot > M0.vs.M1.v1.svg", ignore.stderr = T)
```

```{r echo=FALSE}
knitr::include_graphics("M0.vs.M1.v1.svg")
```

The other way is to collapse them into one vertex with `collapseAtomsIntoMetabolites`:

```{r}
m2 <- collapseAtomsIntoMetabolites(m.ext)
saveModuleToDot(m2, file="M0.vs.M1.v2.dot", name="M0.vs.M1")
system("neato -Tsvg M0.vs.M1.v2.dot > M0.vs.M1.v2.svg", ignore.stderr = T)
```

```{r echo=FALSE}
knitr::include_graphics("M0.vs.M1.v2.svg")
```

# Example on full data

Pre-generated networks and annotations can be downloaded from http://artyomovlab.wustl.edu/publications/supp_materials/GATOM/:

```{r}
network <- readRDS(url("http://artyomovlab.wustl.edu/publications/supp_materials/GATOM/network.kegg.rds"))
org.Mm.eg.gatom.anno <- readRDS(url("http://artyomovlab.wustl.edu/publications/supp_materials/GATOM/org.Mm.eg.gatom.anno.rds"))
# org.Hs.eg.gatom.anno <- readRDS(url("http://artyomovlab.wustl.edu/publications/supp_materials/GATOM/org.Hs.eg.gatom.anno.rds"))
met.db <- readRDS(url("http://artyomovlab.wustl.edu/publications/supp_materials/GATOM/met.kegg.db.rds"))
```

Full dataset for the example can be downloaded here:

```{r message=FALSE}
library(R.utils)
library(data.table)
met.de.raw <- fread("http://artyomovlab.wustl.edu/publications/supp_materials/GAM/Ctrl.vs.MandLPSandIFNg.met.de.tsv.gz")

gene.de.raw <- fread("http://artyomovlab.wustl.edu/publications/supp_materials/GAM/Ctrl.vs.MandLPSandIFNg.gene.de.tsv.gz")
```

And run `gatom` on this dataset:

```{r}
g <- makeMetabolicGraph(network=network, 
                        topology="atoms",
                        org.gatom.anno=org.Mm.eg.gatom.anno, 
                        gene.de=gene.de.raw,
                        met.db=met.db, 
                        met.de=met.de.raw)

gs <- scoreGraph(g, k.gene = 50, k.met = 50)

solver <- rnc_solver()
set.seed(42)
sol <- solve_mwcsp(solver, gs) 
m <- sol$graph
m
```

# Misc

## Using exact solver

For proper analysis quality we recommend to use exact SGMWCS solver `virgo_solver()`, which requires Java (version >= 11) and CPLEX (version >= 12.7) to be installed. If the requirements are met
you can then find a module as following:

```{r eval=FALSE}
vsolver <- virgo_solver(cplex_dir=Sys.getenv("CPLEX_HOME"), threads=4, penalty=0.001)
sol <- solve_mwcsp(vsolver, gs) 
m <- sol$graph
```

Edge penalty option there is used to remove excessive redundancy in genes.

## Pathway annotation

To get functional annotation of obtained modules by KEGG and Reactome
metabolic pathways, we can use hypergeometric test with `fora()` function from `fgsea` package.

```{r}
foraRes <- fgsea::fora(pathways=org.Mm.eg.gatom.anno$pathways,
                       genes=E(m)$gene,
                       universe=unique(E(g)$gene),
                       minSize=5)
foraRes[padj < 0.05]
```

Optionally, redundancy in pathways can be decreased with `collapsePathwaysORA()` function:

```{r}
mainPathways <- fgsea::collapsePathwaysORA(
  foraRes[padj < 0.05],
  pathways=org.Mm.eg.gatom.anno$pathways,
  genes=E(m)$gene,
  universe=unique(E(g)$gene))
foraRes[pathway %in% mainPathways$mainPathways]
```

## Running with no metabolite data

If there is no metabolite data in your experiment assign `met.de` and `k.met` to `NULL`:

```{r}
g <- makeMetabolicGraph(network=networkEx, 
                        topology="atoms",
                        org.gatom.anno=org.Mm.eg.gatom.annoEx, 
                        gene.de=gene.de.rawEx,
                        met.db=met.kegg.dbEx, 
                        met.de=NULL)
gs <- scoreGraph(g, k.gene = 50, k.met = NULL)
```

## Using metabolite-level network

Sometime it could make sense to work with metabolite-metabolite topology of the network, not atom-atom one. Such network is less structured, but contain more genes.


```{r}
gm <- makeMetabolicGraph(network=network, 
                        topology="metabolite",
                        org.gatom.anno=org.Mm.eg.gatom.anno, 
                        gene.de=gene.de.raw,
                        met.db=met.db, 
                        met.de=met.de.raw)

gms <- scoreGraph(gm, k.gene = 50, k.met = 50)

solver <- rnc_solver()
set.seed(42)
sol <- solve_mwcsp(solver, gms) 
mm <- sol$graph
mm
```
